# Network Utility Tools 
- Tools used for network management, routing, scanning and discovery, etc.
    - ping
    - host
    - dig
    - route
    - dhclient
    - arp
    - netcat
    - nmap
    
## Note: start with ! to run bash commands

## Install net-tools for many network utilities

In [1]:
! echo kali | sudo -S apt install net-tools

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
net-tools is already the newest version (2.10-0.1).
net-tools set to manually installed.
The following packages were automatically installed and are no longer required:
  bluez-firmware catfish dh-elpa-helper docutils-common firmware-ath9k-htc
  firmware-atheros firmware-brcm80211 firmware-intel-sound firmware-iwlwifi
  firmware-libertas firmware-realtek firmware-sof-signed
  firmware-ti-connectivity firmware-zd1211 gir1.2-gtksource-3.0
  gir1.2-javascriptcoregtk-4.0 gir1.2-soup-2.4 gir1.2-webkit2-4.0
  gir1.2-xfconf-0 gobject-introspection kali-linux-firmware king-phisher
  libblockdev-crypto2 libblockdev-fs2 libblockdev-loop2 libblockdev-part-err2
  libblockdev-part2 libblockdev-swap2 libblockdev-utils2 libblockdev2
  libcfitsio9 libgdal31 libgdal32 libgeos3.11.1 libmongocrypt0 libmpdec3
  libmujs2 libncurses5 libnginx-mod-http-geoip libnginx-mod-http-image-filter
  libnginx-mod-http-xslt

## ping - test if a system is up

- check if there's a connectivity between the hosts

In [3]:
! ping google.com
# doesn't stop; interrupt the control to send Ctrol+C command to stop pinging

PING google.com (142.250.72.14) 56(84) bytes of data.
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=1 ttl=128 time=23.7 ms
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=2 ttl=128 time=13.0 ms
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=3 ttl=128 time=9.15 ms
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=4 ttl=128 time=20.1 ms
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=5 ttl=128 time=17.7 ms
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=7 ttl=128 time=28.9 ms
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=8 ttl=128 time=11.2 ms
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=9 ttl=128 time=28.0 ms
^C

--- google.com ping statistics ---
9 packets transmitted, 8 received, 11.1111% packet loss, time 8019ms
rtt min/avg/max/mdev = 9.146/18.957/28.901/7.064 ms


In [4]:
! man ping

PING(8)                             iputils                            PING(8)

NAME
       ping - send ICMP ECHO_REQUEST to network hosts

SYNOPSIS

       ping [-aAbBdCDefhLnOqrRUvV46] [-c count] [-F flowlabel] [-i interval]
            [-I interface] [-l preload] [-m mark] [-M pmtudisc_option]
            [-N nodeinfo_option] [-w deadline] [-W timeout] [-p pattern]
            [-Q tos] [-s packetsize] [-S sndbuf] [-t ttl]
            [-T timestamp option] [hop...] {destination}

DESCRIPTION
       ping uses the ICMP protocol's mandatory ECHO_REQUEST datagram to elicit
       an ICMP ECHO_RESPONSE from a host or gateway. ECHO_REQUEST datagrams
       (“pings”) have an IP and ICMP header, followed by a struct timeval and
       then an arbitrary number of “pad” bytes used to fill out the packet.

       ping works with both IPv4 and IPv6. Using only one of them explicitly
       can be enforced by specifying -4 or -6.

       ping can also send IPv6 Node Information Queries (RFC4620).

In [5]:
! ping -c 2 google.com
# count for 2 ECHO_RESPONSE packets and stop

PING google.com (142.250.72.14) 56(84) bytes of data.
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=1 ttl=128 time=11.8 ms
64 bytes from den08s06-in-f14.1e100.net (142.250.72.14): icmp_seq=2 ttl=128 time=34.5 ms

--- google.com ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1002ms
rtt min/avg/max/mdev = 11.819/23.170/34.521/11.351 ms


## DNS Lookup Commands

- host and dig are two utility commands that can lookup DNS

In [2]:
! man host

HOST(1)                             BIND 9                             HOST(1)

NAME
       host - DNS lookup utility

SYNOPSIS
       host  [-aACdlnrsTUwv]  [-c  class] [-N ndots] [-p port] [-R number] [-t
       type] [-W wait] [-m flag] [ [-4] | [-6] ] [-v] [-V] {name} [server]

DESCRIPTION
       host is a simple utility for performing DNS  lookups.  It  is  normally
       used to convert names to IP addresses and vice versa. When no arguments
       or options are given, host prints a short summary of its  command  line
       arguments and options.

       name  is the domain name that is to be looked up. It can also be a dot‐
       ted-decimal IPv4 address or a colon-delimited IPv6  address,  in  which
       case  host  will  by default perform a reverse lookup for that address.
       server is an optional argument which is either the name or  IP  address
       of  the  name  server  that  host should query instead of the server or
       servers listed in /etc/resolv.conf.

In [6]:
! host google.com

google.com has address 142.250.69.238
google.com has IPv6 address 2607:f8b0:400f:803::200e
google.com mail is handled by 10 smtp.google.com.


In [7]:
! host -a google.com # display all records

Trying "google.com"
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 42293
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;google.com.			IN	ANY

;; ANSWER SECTION:
google.com.		5	IN	A	142.250.69.238
google.com.		5	IN	AAAA	2607:f8b0:400f:803::200e
google.com.		5	IN	MX	10 smtp.google.com.

Received 93 bytes from 192.168.47.2#53 in 12 ms


In [8]:
# -t: query type; a
! host -t a coloradomesa.edu # domain ip

coloradomesa.edu has address 10.0.2.54


In [9]:
! host -t mx coloradomesa.edu # domain mail server

coloradomesa.edu mail is handled by 10 mail.coloradomesa.edu.


In [7]:
! host -t ns coloradomesa.edu # nameserver

coloradomesa.edu name server ns2.coloradomesa.edu.
coloradomesa.edu name server ns1.coloradomesa.edu.


## reverse lookup
- given IP, find the domain name

In [10]:
! host 72.166.224.167

167.224.166.72.in-addr.arpa domain name pointer coloradomesa.edu.


In [11]:
# dig is another tool
! man dig

DIG(1)                              BIND 9                              DIG(1)

NAME
       dig - DNS lookup utility

SYNOPSIS
       dig  [@server] [-b address] [-c class] [-f filename] [-k filename] [-m]
       [-p port#] [-q name] [-t type] [-v] [-x addr]  [-y  [hmac:]name:key]  [
       [-4] | [-6] ] [name] [type] [class] [queryopt...]

       dig [-h]

       dig [global-queryopt...] [query...]

DESCRIPTION
       dig  is a flexible tool for interrogating DNS name servers. It performs
       DNS lookups and displays the answers that are returned  from  the  name
       server(s)  that  were queried. Most DNS administrators use dig to trou‐
       bleshoot DNS problems because of its flexibility, ease of use and clar‐
       ity  of output. Other lookup tools tend to have less functionality than
       dig.

       Although dig is normally used with command-line arguments, it also  has
       a  batch  mode  of operation for reading lookup requests from a file. A
       brief summa

In [9]:
! dig google.com


; <<>> DiG 9.16.15-Debian <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 65445
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; MBZ: 0x0005, udp: 4096
;; QUESTION SECTION:
;google.com.			IN	A

;; ANSWER SECTION:
google.com.		5	IN	A	142.250.125.100
google.com.		5	IN	A	142.250.125.102
google.com.		5	IN	A	142.250.125.139
google.com.		5	IN	A	142.250.125.101
google.com.		5	IN	A	142.250.125.138
google.com.		5	IN	A	142.250.125.113

;; Query time: 16 msec
;; SERVER: 192.168.195.2#53(192.168.195.2)
;; WHEN: Wed Sep 08 11:47:12 MDT 2021
;; MSG SIZE  rcvd: 135



## Display arp/routing cache

In [13]:
! man route

ROUTE(8)              Linux System Administrator's Manual             ROUTE(8)

NAME
       route - show / manipulate the IP routing table

SYNOPSIS
       route [-CFvnNee] [-A family |-4|-6]

       route  [-v] [-A family |-4|-6] add [-net|-host] target [netmask Nm] [gw
              Gw] [metric N] [mss M] [window W] [irtt I] [reject] [mod]  [dyn]
              [reinstate] [[dev] If]

       route  [-v] [-A family |-4|-6] del [-net|-host] target [gw Gw] [netmask
              Nm] [metric M] [[dev] If]

       route  [-V] [--version] [-h] [--help]

DESCRIPTION
       Route manipulates the kernel's IP routing tables.  Its primary  use  is
       to  set up static routes to specific hosts or networks via an interface
       after it has been configured with the ifconfig(8) program.

       When the add or del options are used, route modifies  the  routing  ta‐
       bles.   Without  these  options, route displays the current contents of
       the routing tables.

OPTIONS
       -A fami

In [11]:
# route is priviledge command
! echo kali | sudo -S route -n # show numerical addresses

[sudo] password for kali: Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         192.168.47.2    0.0.0.0         UG    100    0        0 eth0
192.168.47.0    0.0.0.0         255.255.255.0   U     100    0        0 eth0


In [12]:
! echo kali | sudo -S ifconfig

[sudo] password for kali: docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether 02:42:df:eb:f1:2a  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.47.130  netmask 255.255.255.0  broadcast 192.168.47.255
        inet6 fe80::ed5d:b4c2:d275:17ca  prefixlen 64  scopeid 0x20<link>
        ether 00:0c:29:8a:20:62  txqueuelen 1000  (Ethernet)
        RX packets 463403  bytes 683071714 (651.4 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 34114  bytes 2680677 (2.5 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000 

In [13]:
! echo kali | sudo -S route -n # show numerical addresses

[sudo] password for kali: Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         192.168.195.2   0.0.0.0         UG    100    0        0 eth0
192.168.195.0   0.0.0.0         255.255.255.0   U     100    0        0 eth0


## dhclient
 - Dynamic Host Configuration Prototocol Client
 - used to configure network interfaces with DHCP
     - IP addresses are assigned dynamically
     - if fails, IP can be assigned statically

In [14]:
! man dhclient

dhclient(8)                 System Manager's Manual                dhclient(8)

NAME
       dhclient - Dynamic Host Configuration Protocol Client

SYNOPSIS
       dhclient  [ -4 | -6 ] [ -S ] [ -N [ -N...  ] ] [ -T [ -T...  ] ] [ -P [
       -P...  ] ] -R ] [ -i ] [ -I ] [ -4o6 port ] [ -D LL|LLT ]  [  -p  port-
       number  ] [ -d ] [ -df duid-lease-file ] [ -e VAR=value ] [ -q ] [ -1 ]
       [ -r | -x ] [ -lf lease-file ] [ -pf pid-file ] [ --no-pid ] [ -cf con‐
       fig-file ] [ -sf script-file ] [ -s server-addr ] [ -g relay ] [ -n ] [
       -nw ] [ -w ] [ --dad-wait-time seconds ] [ --prefix-len-hint length ] [
       --decline-wait-time seconds ] [ -v ] [ --version ] [ if0 [ ...ifN ] ]

DESCRIPTION
       The Internet Systems Consortium DHCP Client, dhclient, provides a means
       for configuring one or more network interfaces using the  Dynamic  Host
       Configuration  Protocol, BOOTP protocol, or if these protocols fail, by
       statically assigning an address.

OP

In [15]:
! echo kali | sudo -S dhclient -v -r

[sudo] password for kali: Killed old client process
Internet Systems Consortium DHCP Client 4.4.1
Copyright 2004-2018 Internet Systems Consortium.
All rights reserved.
For info, please visit https://www.isc.org/software/dhcp/

Listening on LPF/eth0/00:0c:29:17:1a:37
Sending on   LPF/eth0/00:0c:29:17:1a:37
Sending on   Socket/fallback
DHCPRELEASE of 192.168.195.164 on eth0 to 192.168.195.254 port 67


## Internal routing table
- typically displays gateway information that could be very useful to know how the packet is being routed to the internet

In [19]:
! echo kali | sudo -S dhclient

[sudo] password for kali: RTNETLINK answers: File exists


In [20]:
! route -n

Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         192.168.195.2   0.0.0.0         UG    0      0        0 eth0
192.168.195.0   0.0.0.0         255.255.255.0   U     0      0        0 eth0


In [21]:
! echo kali | sudo -S ifconfig

[sudo] password for kali: eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.195.164  netmask 255.255.255.0  broadcast 192.168.195.255
        inet6 fe80::20c:29ff:fe17:1a37  prefixlen 64  scopeid 0x20<link>
        ether 00:0c:29:17:1a:37  txqueuelen 1000  (Ethernet)
        RX packets 466112  bytes 661770070 (631.1 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 152785  bytes 10730159 (10.2 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 180590  bytes 64790961 (61.7 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 180590  bytes 64790961 (61.7 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0



## ARP - address resolution protocol 
- maps private IP to ethernet/MAC address for local/internal routing
- without knowing each other's MAC address, systems can't communicate with each other
- there are 4 different types of ARP messages, but ARP request and ARP reply message are most important and common types

In [22]:
! man arp

ARP(8)                Linux System Administrator's Manual               ARP(8)

NAME
       arp - manipulate the system ARP cache

SYNOPSIS
       arp [-vn] [-H type] [-i if] [-ae] [hostname]

       arp [-v] [-i if] -d hostname [pub]

       arp [-v] [-H type] [-i if] -s hostname hw_addr [temp]

       arp [-v] [-H type] [-i if] -s hostname hw_addr [netmask nm] pub

       arp [-v] [-H type] [-i if] -Ds hostname ifname [netmask nm] pub

       arp [-vnD] [-H type] [-i if] -f [filename]

DESCRIPTION
       Arp  manipulates or displays the kernel's IPv4 network neighbour cache.
       It can add entries to the table, delete one or display the current con‐
       tent.

       ARP  stands  for Address Resolution Protocol, which is used to find the
       media access control address of a network neighbour for  a  given  IPv4
       Address.

MODES
       arp with no mode specifier will print the current content of the table.
       It is possible to limit the number of entries printed, b

In [17]:
! echo kali | sudo -S arp -n # -n shows numerical addresses from ARP cache

[sudo] password for kali: Address                  HWtype  HWaddress           Flags Mask            Iface
192.168.47.129           ether   00:0c:29:b0:50:a7   C                     eth0
192.168.47.254           ether   00:50:56:f7:81:3e   C                     eth0
192.168.47.2             ether   00:50:56:f6:54:9a   C                     eth0


### Find ip of host or another VM or gateway and ping it to see the ARP cache being changed
- $ ifconfig | grep inet 
- and ping host and then run route -n again to see the routing table

In [16]:
! ping -c 2 192.168.47.129

PING 192.168.47.129 (192.168.47.129) 56(84) bytes of data.
64 bytes from 192.168.47.129: icmp_seq=1 ttl=64 time=0.888 ms
64 bytes from 192.168.47.129: icmp_seq=2 ttl=64 time=0.557 ms

--- 192.168.47.129 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1009ms
rtt min/avg/max/mdev = 0.557/0.722/0.888/0.165 ms


In [25]:
! ping -c 1 -w 1 192.168.195.131

PING 192.168.195.131 (192.168.195.131) 56(84) bytes of data.

--- 192.168.195.131 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms



In [26]:
! ping -c 1 192.168.231.155

PING 192.168.231.155 (192.168.231.155) 56(84) bytes of data.

--- 192.168.231.155 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms



In [19]:
! echo kali | sudo -S arp -n

[sudo] password for kali: Address                  HWtype  HWaddress           Flags Mask            Iface
192.168.47.2             ether   00:50:56:f6:54:9a   C                     eth0


In [28]:
# arp -d <ip> to delete entry
! echo kali | sudo -S arp -d 192.168.195.131

[sudo] password for kali: 

In [29]:
! echo kali | sudo -S arp -n #shows (incomplete) completely clearing is not easy

[sudo] password for kali: Address                  HWtype  HWaddress           Flags Mask            Iface
192.168.195.2            ether   00:50:56:e4:5b:86   C                     eth0


### flush all entries

In [18]:
! echo kali | sudo -S ip -s -s neigh flush all

[sudo] password for kali: 192.168.47.129 dev eth0 lladdr 00:0c:29:b0:50:a7  used 120/120/100probes 4 STALE 
192.168.47.254 dev eth0 lladdr 00:50:56:f7:81:3e  used 509/509/477probes 1 STALE 
192.168.47.2 dev eth0 lladdr 00:50:56:f6:54:9a  used 176/176/155probes 1 STALE 

*** Round 1, deleting 3 entries ***
*** Flush is complete after 1 round ***


## Netcat
- utility tool that supports a wide range of commands to manage networks and send and receive data between systems
- more info: https://www.varonis.com/blog/netcat-commands/
- netcat or nc are the commands used with many options

In [47]:
! nc -h

[v1.10-45]
connect to somewhere:	nc [-options] hostname port[s] [ports] ... 
listen for inbound:	nc -l -p port [-options] [hostname] [port]
options:
	-c shell commands	as `-e'; use /bin/sh to exec [dangerous!!]
	-e filename		program to exec after connect [dangerous!!]
	-b			allow broadcasts
	-g gateway		source-routing hop point[s], up to 8
	-G num			source-routing pointer: 4, 8, 12, ...
	-h			this cruft
	-i secs			delay interval for lines sent, ports scanned
        -k                      set keepalive option on socket
	-l			listen mode, for inbound connects
	-n			numeric-only IP addresses, no DNS
	-o file			hex dump of traffic
	-p port			local port number
	-r			randomize local and remote ports
	-q secs			quit after EOF on stdin and delay of secs
	-s addr			local source address
	-T tos			set Type Of Service
	-t			answer TELNET negotiation
	-u			UDP mode
	-v			verbose [use twice to be more verbose]
	-w secs			timeout for connects and final net reads
	-C			Send CRLF as line-ending
	-z		

### Port scanning with Netcat
- netcat can be used scan a host(domain or ip) to diagnose a network issues or performance problem
- can determine whether a firewall is blocking traffic
- didn't find a success!

In [55]:
! echo kali | sudo -S ifconfig | grep inet

[sudo] password for kali:         inet 192.168.195.160  netmask 255.255.255.0  broadcast 192.168.195.255
        inet6 fe80::20c:29ff:febb:4e8a  prefixlen 64  scopeid 0x20<link>
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>


In [56]:
! ping -c 1 google.com

PING google.com (172.217.12.14) 56(84) bytes of data.
64 bytes from den02s02-in-f14.1e100.net (172.217.12.14): icmp_seq=1 ttl=128 time=11.0 ms

--- google.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 11.040/11.040/11.040/0.000 ms


In [ ]:
# scan port from 1-1000; domain name isn't resolved correctly, use IP
! nc -v -z -n -w1 142.250.69.238 443 1-1000

(UNKNOWN) [142.250.69.238] 443 (https) open
(UNKNOWN) [142.250.69.238] 1000 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 999 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 998 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 997 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 996 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 995 (pop3s) : Connection timed out
(UNKNOWN) [142.250.69.238] 994 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 993 (imaps) : Connection timed out
(UNKNOWN) [142.250.69.238] 992 (telnets) : Connection timed out
(UNKNOWN) [142.250.69.238] 991 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 990 (ftps) : Connection timed out
(UNKNOWN) [142.250.69.238] 989 (ftps-data) : Connection timed out
(UNKNOWN) [142.250.69.238] 988 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 987 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 986 (?) : Connection timed out
(UNKNOWN) [142.250.69.238] 985 (?) : Connection timed out
(U

In [41]:
# check single port
! nc -z -v google.com 80

DNS fwd/rev mismatch: google.com != den08s06-in-f14.1e100.net
google.com [142.250.72.14] 80 (http) open


### Sending data over network
- netcat can connect to udp and tcp service and send data

In [29]:
# download index page from washing.edu and save it to file
# send HTTP GET request
! printf "GET / HTTP/1.1\r\n\r\n" | nc example.com 80 > example.html

In [31]:
! cat example.html

HTTP/1.1 400 Bad Request
Content-Type: text/html
Content-Length: 349
Connection: close
Date: Mon, 28 Aug 2023 17:42:18 GMT
Server: ECSF (dna/63B9)

<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
	<head>
		<title>400 - Bad Request</title>
	</head>
	<body>
		<h1>400 - Bad Request</h1>
	</body>
</html>


### Create a local Web server
- run the following command from a terminal (server)

In [39]:
! nc -l -p 8080
# open a browser and go to localhost:8080

GET / HTTP/1.1
Host: localhost:8080
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
Cookie: username-localhost-8888="2|1:0|10:1693408261|23:username-localhost-8888|44:ZjJiZDBjYTY0NDAxNGNhMmIxNWQyYWExNjE4ZjMyZjY=|c44fc352e07019441b4e5243ac806c2563cc22acdac3f584faefb0001ee1158b"; _xsrf=2|bda8b308|946cc4fe1602258f52761f6acdd53fac|1692630366
Upgrade-Insecure-Requests: 1
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: none
Sec-Fetch-User: ?1

^C



In [38]:
! printf "HTTP/1.1 200 OK \n\n <h1>Welcome...<h1>\r\n\r\n" | netcat -l -p 8080
# open browswer and point it to: localhost:8080
# stop the server or interrupt the kernel to actually send the response

GET / HTTP/1.1
Host: localhost:8080
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
Cookie: username-localhost-8888="2|1:0|10:1693408209|23:username-localhost-8888|44:MzUzODY0NzkxOTZjNDkxMThjNjg5YTgyMmZkMzM1MGM=|6d6cfa644403bf4cf8c1a21d6e31b9f0a16dc68d7c9f81ebd80189a3cb40dbe6"; _xsrf=2|bda8b308|946cc4fe1602258f52761f6acdd53fac|1692630366
Upgrade-Insecure-Requests: 1
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: none
Sec-Fetch-User: ?1

^C



### Netcat File Transfer

### server
$ nc -l [port] > file_name.in # receive a file

### client
$ nc [host] [port] < file_name.out # send a file


### Grab banner

In [53]:
! echo "" | nc -v -w1 coloradomesa.edu 80

DNS fwd/rev mismatch: coloradomesa.edu != haproxy.coloradomesa.edu
coloradomesa.edu [10.0.2.54] 80 (http) open


In [54]:
# use google's IP address
! echo "" | nc -v -n -w3 142.250.69.238 443
# -v - verbose
# -n don't resolve dns
# -w1 wait for 1 second for the response

(UNKNOWN) [142.250.69.238] 443 (https) open


### Netcat Backdoor Shells

```bash
nc -l -p [port] -e /bin/bash # run a shell on Linux
nc -l -p [port] -e cmd.exe # run a shell on Netcat for Windows
```
- run nc on a terminal and connect to it with another

## Network scaning and mapping

- nmap is the most popular tool for this - https://nmap.org/

In [54]:
! nmap --help

Nmap 7.80 ( https://nmap.org )
Usage: nmap [Scan Type(s)] [Options] {target specification}
TARGET SPECIFICATION:
  Can pass hostnames, IP addresses, networks, etc.
  Ex: scanme.nmap.org, microsoft.com/24, 192.168.0.1; 10.0.0-255.1-254
  -iL <inputfilename>: Input from list of hosts/networks
  -iR <num hosts>: Choose random targets
  --exclude <host1[,host2][,host3],...>: Exclude hosts/networks
  --excludefile <exclude_file>: Exclude list from file
HOST DISCOVERY:
  -sL: List Scan - simply list targets to scan
  -sn: Ping Scan - disable port scan
  -Pn: Treat all hosts as online -- skip host discovery
  -PS/PA/PU/PY[portlist]: TCP SYN/ACK, UDP or SCTP discovery to given ports
  -PE/PP/PM: ICMP echo, timestamp, and netmask request discovery probes
  -PO[protocol list]: IP Protocol Ping
  -n/-R: Never do DNS resolution/Always resolve [default: sometimes]
  --dns-servers <serv1[,serv2],...>: Specify custom DNS servers
  --system-dns: Use OS's DNS resolver
  --traceroute: Trace hop path to 

In [32]:
! nmap -v -A scanme.nmap.org

Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 08:32 MDT
NSE: Loaded 153 scripts for scanning.
NSE: Script Pre-scanning.
Initiating NSE at 08:32
Completed NSE at 08:32, 0.00s elapsed
Initiating NSE at 08:32
Completed NSE at 08:32, 0.00s elapsed
Initiating NSE at 08:32
Completed NSE at 08:32, 0.00s elapsed
Initiating Ping Scan at 08:32
Scanning scanme.nmap.org (45.33.32.156) [2 ports]
Completed Ping Scan at 08:32, 0.01s elapsed (1 total hosts)
Initiating Parallel DNS resolution of 1 host. at 08:32
Completed Parallel DNS resolution of 1 host. at 08:33, 1.04s elapsed
Initiating Connect Scan at 08:33
Scanning scanme.nmap.org (45.33.32.156) [1000 ports]
Discovered open port 443/tcp on 45.33.32.156
Discovered open port 80/tcp on 45.33.32.156
Discovered open port 53/tcp on 45.33.32.156
Discovered open port 22/tcp on 45.33.32.156
Discovered open port 8010/tcp on 45.33.32.156
Discovered open port 31337/tcp on 45.33.32.156
Discovered open port 8008/tcp on 45.33.32.156
Discovered open port

### Scanning unauthorized system and network is illegal!

### Scan Metasploitable VM
- find IP address of the VM
- detect OS and services

In [64]:
! nmap -A 192.168.195.131

Starting Nmap 7.80 ( https://nmap.org ) at 2020-09-14 16:16 EDT
Nmap scan report for 192.168.195.131
Host is up (0.00069s latency).
Not shown: 994 closed ports
PORT     STATE SERVICE    VERSION
21/tcp   open  ftp        vsftpd 3.0.3
22/tcp   open  ssh        OpenSSH 7.2p2 Ubuntu 4ubuntu2.2 (Ubuntu Linux; protocol 2.0)
| ssh-hostkey: 
|   2048 4b:6f:e9:4a:8e:5b:2b:4d:12:34:94:48:9b:fc:05:1a (RSA)
|   256 37:61:b8:e9:07:af:1c:f1:6a:49:94:ea:de:19:cf:b4 (ECDSA)
|_  256 5c:c2:48:7e:04:51:1d:ba:89:a1:ca:4a:55:90:4f:e6 (ED25519)
23/tcp   open  telnet     Linux telnetd
53/tcp   open  domain     ISC BIND 9.10.3-P4 (Ubuntu Linux)
| dns-nsid: 
|_  bind.version: 9.10.3-P4-Ubuntu
80/tcp   open  http       Apache httpd 2.4.18 ((Ubuntu))
|_http-server-header: Apache/2.4.18 (Ubuntu)
|_http-title: Apache2 Ubuntu Default Page: It works
3128/tcp open  http-proxy Squid http proxy 3.5.12
|_http-server-header: squid/3.5.12
|_http-title: ERROR: The requested URL could not be retrieved
Service Info: OSs: Uni

### Nmap Common Commands
- https://securitytrails.com/blog/top-15-nmap-commands-to-scan-remote-hosts

### Basic Namp Scan against IP or host
- perfect first step quick scanning

In [65]:
! nmap 192.168.195.131

Starting Nmap 7.80 ( https://nmap.org ) at 2020-09-14 16:18 EDT
Nmap scan report for 192.168.195.131
Host is up (0.0013s latency).
Not shown: 994 closed ports
PORT     STATE SERVICE
21/tcp   open  ftp
22/tcp   open  ssh
23/tcp   open  telnet
53/tcp   open  domain
80/tcp   open  http
3128/tcp open  squid-http

Nmap done: 1 IP address (1 host up) scanned in 0.08 seconds


In [66]:
! nmap scanme.nmap.org

Starting Nmap 7.80 ( https://nmap.org ) at 2020-09-14 16:19 EDT
Nmap scan report for scanme.nmap.org (45.33.32.156)
Host is up (0.065s latency).
Other addresses for scanme.nmap.org (not scanned): 2600:3c01::f03c:91ff:fe18:bb2f
Not shown: 988 closed ports
PORT      STATE    SERVICE
22/tcp    open     ssh
25/tcp    filtered smtp
53/tcp    open     domain
80/tcp    open     http
135/tcp   filtered msrpc
139/tcp   filtered netbios-ssn
443/tcp   open     https
445/tcp   filtered microsoft-ds
8008/tcp  open     http
8010/tcp  open     xmpp
9929/tcp  open     nping-echo
31337/tcp open     Elite

Nmap done: 1 IP address (1 host up) scanned in 2.55 seconds


### Scan specific ports
- nmap can scan all 65535 possible ports
- you can limit to specific ports

In [33]:
! nmap -p 1-65535 localhost

Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 08:33 MDT
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000077s latency).
Other addresses for localhost (not scanned): ::1
Not shown: 65491 closed ports
PORT      STATE SERVICE
3780/tcp  open  nnp
8888/tcp  open  sun-answerbook
8889/tcp  open  ddi-tcp-2
33509/tcp open  unknown
34037/tcp open  unknown
34421/tcp open  unknown
34765/tcp open  unknown
35473/tcp open  unknown
35583/tcp open  unknown
36855/tcp open  unknown
36919/tcp open  unknown
37089/tcp open  unknown
37475/tcp open  neckar
37607/tcp open  unknown
37895/tcp open  unknown
38909/tcp open  unknown
39799/tcp open  unknown
40313/tcp open  unknown
40419/tcp open  unknown
42735/tcp open  unknown
43431/tcp open  unknown
43821/tcp open  unknown
44399/tcp open  unknown
44671/tcp open  unknown
45343/tcp open  unknown
46037/tcp open  unknown
46137/tcp open  unknown
46267/tcp open  unknown
46827/tcp open  unknown
47061/tcp open  unknown
47371/tcp open  unknown
47827/tcp 

In [67]:
# scan Metasploitable VM - find it's IP first
! nmap -p 22,80,443 192.168.195.131

Starting Nmap 7.80 ( https://nmap.org ) at 2020-09-14 16:21 EDT
Nmap scan report for 192.168.195.131
Host is up (0.00059s latency).

PORT    STATE  SERVICE
22/tcp  open   ssh
80/tcp  open   http
443/tcp closed https

Nmap done: 1 IP address (1 host up) scanned in 0.04 seconds


### Scan multiple IP addresses
- nmap ip1 ip2
- nmap 1.1.1.1,2,3,4 # consecutive ip addresses

### Scan IP ranges
- Classless Inter-Domain Routing (CIDR) notation: https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing
    - See IPv4 CIDR blocsk section for quick reference
- nmap 192.168.195.0/24 # Scan $2^8=256$ hosts in 195 subnet; class C
- nmap 192.168.195.* # scan all 256 hosts from .1 to .256; class
- nmap 192.168.195.10-100 # scan host from .10 to .100
- nmap 192.168.195.* --exclude 192.168.195.1

### Scan the most popular ports
- scan top N most popular ports

In [68]:
# scan top ports on Metasploitable VM
! nmap --top-ports 10 192.168.195.131

Starting Nmap 7.80 ( https://nmap.org ) at 2020-09-14 16:22 EDT
Nmap scan report for 192.168.195.131
Host is up (0.00058s latency).

PORT     STATE  SERVICE
21/tcp   open   ftp
22/tcp   open   ssh
23/tcp   open   telnet
25/tcp   closed smtp
80/tcp   open   http
110/tcp  closed pop3
139/tcp  closed netbios-ssn
443/tcp  closed https
445/tcp  closed microsoft-ds
3389/tcp closed ms-wbt-server

Nmap done: 1 IP address (1 host up) scanned in 0.04 seconds


### Scan hosts and IP addresses reading from a text file
- create a text file (hosts.txt) with ip address and host names one per line

In [69]:
! printf "192.168.195.131\nscanme.nmap.org" > hosts.txt

In [70]:
! cat hosts.txt

192.168.195.131
scanme.nmap.org

In [71]:
! nmap -iL hosts.txt

Starting Nmap 7.80 ( https://nmap.org ) at 2020-09-14 16:23 EDT
Nmap scan report for 192.168.195.131
Host is up (0.0020s latency).
Not shown: 994 closed ports
PORT     STATE SERVICE
21/tcp   open  ftp
22/tcp   open  ssh
23/tcp   open  telnet
53/tcp   open  domain
80/tcp   open  http
3128/tcp open  squid-http

Nmap scan report for scanme.nmap.org (45.33.32.156)
Host is up (0.061s latency).
Other addresses for scanme.nmap.org (not scanned): 2600:3c01::f03c:91ff:fe18:bb2f
Not shown: 988 closed ports
PORT      STATE    SERVICE
22/tcp    open     ssh
25/tcp    filtered smtp
53/tcp    open     domain
80/tcp    open     http
135/tcp   filtered msrpc
139/tcp   filtered netbios-ssn
443/tcp   open     https
445/tcp   filtered microsoft-ds
8008/tcp  open     http
8010/tcp  open     xmpp
9929/tcp  open     nping-echo
31337/tcp open     Elite

Nmap done: 2 IP addresses (2 hosts up) scanned in 1.75 seconds


### Save scan results to a file
- oN - normal format
- oX - xml format

In [72]:
! nmap -oN nmapResults.txt localhost

Starting Nmap 7.80 ( https://nmap.org ) at 2020-09-14 16:24 EDT
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000052s latency).
Other addresses for localhost (not scanned): ::1
Not shown: 999 closed ports
PORT     STATE SERVICE
8888/tcp open  sun-answerbook

Nmap done: 1 IP address (1 host up) scanned in 0.04 seconds


In [73]:
! cat nmapResults.txt

# Nmap 7.80 scan initiated Mon Sep 14 16:24:12 2020 as: nmap -oN nmapResults.txt localhost
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000052s latency).
Other addresses for localhost (not scanned): ::1
Not shown: 999 closed ports
PORT     STATE SERVICE
8888/tcp open  sun-answerbook

# Nmap done at Mon Sep 14 16:24:12 2020 -- 1 IP address (1 host up) scanned in 0.04 seconds


In [34]:
! nmap -oX nmapResults.xml localhost

Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 08:35 MDT
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000061s latency).
Other addresses for localhost (not scanned): ::1
Not shown: 999 closed ports
PORT     STATE SERVICE
8888/tcp open  sun-answerbook

Nmap done: 1 IP address (1 host up) scanned in 0.07 seconds


In [35]:
! cat nmapResults.xml

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE nmaprun>
<?xml-stylesheet href="file:///usr/bin/../share/nmap/nmap.xsl" type="text/xsl"?>
<!-- Nmap 7.91 scan initiated Fri Sep 10 08:35:27 2021 as: nmap -oX nmapResults.xml localhost -->
<nmaprun scanner="nmap" args="nmap -oX nmapResults.xml localhost" start="1631284527" startstr="Fri Sep 10 08:35:27 2021" version="7.91" xmloutputversion="1.05">
<scaninfo type="connect" protocol="tcp" numservices="1000" services="1,3-4,6-7,9,13,17,19-26,30,32-33,37,42-43,49,53,70,79-85,88-90,99-100,106,109-111,113,119,125,135,139,143-144,146,161,163,179,199,211-212,222,254-256,259,264,280,301,306,311,340,366,389,406-407,416-417,425,427,443-445,458,464-465,481,497,500,512-515,524,541,543-545,548,554-555,563,587,593,616-617,625,631,636,646,648,666-668,683,687,691,700,705,711,714,720,722,726,749,765,777,783,787,800-801,808,843,873,880,888,898,900-903,911-912,981,987,990,992-993,995,999-1002,1007,1009-1011,1021-1100,1102,1104-1108,1110-1114,1117,1119,1121-1

### Scan + OS and service detection with fast execution
- T4 - for faster execution

In [36]:
# scan Metasploitable VM
! nmap -A -T4 192.168.195.134

Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 08:36 MDT
Nmap scan report for 192.168.195.134
Host is up (0.0011s latency).
Not shown: 977 closed ports
PORT     STATE SERVICE     VERSION
21/tcp   open  ftp         vsftpd 2.3.4
|_ftp-anon: Anonymous FTP login allowed (FTP code 230)
| ftp-syst: 
|   STAT: 
| FTP server status:
|      Connected to 192.168.195.164
|      Logged in as ftp
|      TYPE: ASCII
|      No session bandwidth limit
|      Session timeout in seconds is 300
|      Control connection is plain text
|      Data connections will be plain text
|      vsFTPd 2.3.4 - secure, fast, stable
|_End of status
22/tcp   open  ssh         OpenSSH 4.7p1 Debian 8ubuntu1 (protocol 2.0)
| ssh-hostkey: 
|   1024 60:0f:cf:e1:c0:5f:6a:74:d6:90:24:fa:c4:d5:6c:cd (DSA)
|_  2048 56:56:24:0f:21:1d:de:a7:2b:ae:61:b1:24:3d:e8:f3 (RSA)
23/tcp   open  telnet      Linux telnetd
25/tcp   open  smtp        Postfix smtpd
|_smtp-commands: metasploitable.localdomain, PIPELINING, SIZE 10240000, VR

### Detect Service/daemon version
- -sV parameter does it!

In [37]:
! nmap -sV 192.168.195.134

Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 08:39 MDT
Nmap scan report for 192.168.195.134
Host is up (0.0021s latency).
Not shown: 977 closed ports
PORT     STATE SERVICE     VERSION
21/tcp   open  ftp         vsftpd 2.3.4
22/tcp   open  ssh         OpenSSH 4.7p1 Debian 8ubuntu1 (protocol 2.0)
23/tcp   open  telnet      Linux telnetd
25/tcp   open  smtp        Postfix smtpd
53/tcp   open  domain      ISC BIND 9.4.2
80/tcp   open  http        Apache httpd 2.2.8 ((Ubuntu) DAV/2)
111/tcp  open  rpcbind     2 (RPC #100000)
139/tcp  open  netbios-ssn Samba smbd 3.X - 4.X (workgroup: WORKGROUP)
445/tcp  open  netbios-ssn Samba smbd 3.X - 4.X (workgroup: WORKGROUP)
512/tcp  open  exec        netkit-rsh rexecd
513/tcp  open  login?
514/tcp  open  tcpwrapped
1099/tcp open  java-rmi    GNU Classpath grmiregistry
1524/tcp open  bindshell   Metasploitable root shell
2049/tcp open  nfs         2-4 (RPC #100003)
2121/tcp open  ftp         ProFTPD 1.3.1
3306/tcp open  mysql       MySQL 5.0

### Scan TCP or UDP based services
- -sT - scan TCP-based services (most services use TCP)
- sU - scan UDP-based services

In [39]:
! echo kali | sudo -S nmap --top-ports 10 -sU 192.168.195.134
# a little slower

[sudo] password for kali: Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 08:40 MDT
Nmap scan report for 192.168.195.134
Host is up (0.00061s latency).

PORT     STATE         SERVICE
53/udp   open          domain
67/udp   closed        dhcps
123/udp  closed        ntp
135/udp  closed        msrpc
137/udp  open|filtered netbios-ns
138/udp  open|filtered netbios-dgm
161/udp  closed        snmp
445/udp  closed        microsoft-ds
631/udp  closed        ipp
1434/udp closed        ms-sql-m
MAC Address: 00:0C:29:F8:BE:97 (VMware)

Nmap done: 1 IP address (1 host up) scanned in 2.52 seconds


### CVE detection using Nmap
- Nmap supports Nmap Scripting Engine (NSE)
- Nmap can be scripted using Lua programming language
- can use pre-defined scripts or write your own
- See Nmap Scripting Engine for details: https://nmap.org/book/nse-usage.html

- run a full vulnerability scan on a target

In [40]:
# scan Metasploitable VM and detect CVE vulnerabilities
! nmap -Pn --script vuln 192.168.195.134
# takes a several minutes!

Host discovery disabled (-Pn). All addresses will be marked 'up' and scan times will be slower.
Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 08:40 MDT
Pre-scan script results:
| broadcast-avahi-dos: 
|   Discovered hosts:
|     224.0.0.251
|   After NULL UDP avahi packet DoS (CVE-2011-1002).
|_  Hosts are all up (not vulnerable).
Nmap scan report for 192.168.195.134
Host is up (0.29s latency).
Not shown: 977 closed ports
PORT     STATE SERVICE
21/tcp   open  ftp
|_sslv2-drown: 
22/tcp   open  ssh
23/tcp   open  telnet
25/tcp   open  smtp
| smtp-vuln-cve2010-4344: 
|_  The SMTP server is not Exim: NOT VULNERABLE
| ssl-dh-params: 
|   VULNERABLE:
|   Anonymous Diffie-Hellman Key Exchange MitM Vulnerability
|     State: VULNERABLE
|       Transport Layer Security (TLS) services that use anonymous
|       Diffie-Hellman key exchange only provide protection against passive
|       eavesdropping, and are vulnerable to active man-in-the-middle attacks
|       which could completely c

### Denial of Service (DOS) attack with Nmap

```bash
$ nmap 192.168.1.105 -max-parallelism 800 -Pn --script http-slowloris --script-args http-slowloris.runforever=true
```

### Launching brute force attacks

### WordPress brute force attack

```bash
$ nmap -sV --script http-wordpress-brute --script-args 'userdb=users.txt,passdb=passwds.txt,http-wordpress-brute.hostname=domain.com, http-wordpress-brute.threads=3,brute.firstonly=true' [target]
```

### MS-SQL Brute force attack

```bash
$ nmap -p 1433 --script ms-sql-brute --script-args userdb=customuser.txt, passdb=custompass.txt [target]
```

### FTP Brute force attack

```bash
$ nmap --script ftp-brute -p 21 [target]
```

#### SEED Lab has FTP port 21 open and an account with user:user
- let's brute force it!
- Note - doesn't find defualt account seed:dees - interesting!

```bash

$ nmap --script ftp-brute -p 21 192.168.195.131

Starting Nmap 7.80 ( https://nmap.org ) at 2020-09-14 16:28 EDT
NSE: [ftp-brute] usernames: Time limit 10m00s exceeded.
NSE: [ftp-brute] usernames: Time limit 10m00s exceeded.
NSE: [ftp-brute] passwords: Time limit 10m00s exceeded.
Nmap scan report for 192.168.195.131
Host is up (0.00059s latency).

PORT   STATE SERVICE
21/tcp open  ftp
| ftp-brute: 
|   Accounts: 
|     user:user - Valid credentials
|_  Statistics: Performed 3654 guesses in 601 seconds, average tps: 5.9

Nmap done: 1 IP address (1 host up) scanned in 601.09 seconds
```

In [41]:
# brute force Metasploitable ftp account
! nmap --script ftp-brute -p 21 192.168.195.134
# may take some time!

Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 08:47 MDT
NSE: [ftp-brute] usernames: Time limit 10m00s exceeded.
NSE: [ftp-brute] usernames: Time limit 10m00s exceeded.
NSE: [ftp-brute] passwords: Time limit 10m00s exceeded.
Nmap scan report for 192.168.195.134
Host is up (0.00047s latency).

PORT   STATE SERVICE
21/tcp open  ftp
| ftp-brute: 
|   Accounts: 
|     user:user - Valid credentials
|_  Statistics: Performed 3693 guesses in 602 seconds, average tps: 5.9

Nmap done: 1 IP address (1 host up) scanned in 601.95 seconds


### Detecting malware infection on remote hosts
- nmap may detect malware and backdoors by running extensive tests on popular OS services like:
    - IDENTD, Proftpd, Vsftpd, IRC, SMB, and SMTP
- it also checks Google's Safe Browsing and VirusTotal databases

In [42]:
# detect malware infection on Metasploitable
! nmap -sV --script http-malware-host 192.168.195.134

Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 09:03 MDT
Nmap scan report for 192.168.195.134
Host is up (0.0013s latency).
Not shown: 977 closed ports
PORT     STATE SERVICE     VERSION
21/tcp   open  ftp         vsftpd 2.3.4
22/tcp   open  ssh         OpenSSH 4.7p1 Debian 8ubuntu1 (protocol 2.0)
23/tcp   open  telnet      Linux telnetd
25/tcp   open  smtp        Postfix smtpd
53/tcp   open  domain      ISC BIND 9.4.2
80/tcp   open  http        Apache httpd 2.2.8 ((Ubuntu) DAV/2)
|_http-malware-host: Host appears to be clean
|_http-server-header: Apache/2.2.8 (Ubuntu) DAV/2
111/tcp  open  rpcbind     2 (RPC #100000)
| rpcinfo: 
|   program version    port/proto  service
|   100000  2            111/tcp   rpcbind
|   100000  2            111/udp   rpcbind
|   100003  2,3,4       2049/tcp   nfs
|   100003  2,3,4       2049/udp   nfs
|   100005  1,2,3      33520/udp   mountd
|   100005  1,2,3      57262/tcp   mountd
|   100021  1,3,4      43399/udp   nlockmgr
|   100021  1,3,4    

In [43]:
# use Google's malware check on Metasploitable
! nmap -p 80 --script http-google-malware 192.168.195.134
# runs very quick; can't tell if it's working...

Starting Nmap 7.91 ( https://nmap.org ) at 2021-09-10 09:04 MDT
Nmap scan report for 192.168.195.134
Host is up (0.00063s latency).

PORT   STATE SERVICE
80/tcp open  http

Nmap done: 1 IP address (1 host up) scanned in 0.23 seconds
